controllers/validators/validators.go

```diff
@@ -0,0 +1,37 @@
+package validators
+
+import (
+	"fmt"
+
+	"github.com/blang/semver/v4"
+
+	operatorsv1alpha1 "github.com/operator-framework/operator-controller/api/v1alpha1"
+)
+
+type operatorCRValidatorFunc func(operator *operatorsv1alpha1.Operator) error
+
+// validateSemver validates that the operator's version is a valid SemVer.
+// this validation should already be happening at the CRD level. But, it depends
+// on a regex that could possibly fail to validate a valid SemVer. This is added as an
+// extra measure to ensure a valid spec before the CR is processed for resolution
+func validateSemver(operator *operatorsv1alpha1.Operator) error {
```

Comment:
> If the CRD regex validation fails, will the value even be set?
If I understood the question correctly (and please forgive me if I'm going over stuff you already know), my understanding of the way it works is: there's both client-side and server-side validation of the field value against the regex before it reaches the controller. The intent of putting that the validation is to automatically reject anything that isn't a semver so it doesn't even make it to the controller. But, after looking at the regex, I couldn't get any confidence that it would work 100% of the time. I tried to validate with different negative cases, but I couldn't convince myself that it will definitely catch everything. So, it could be possible that an invalid spec slips through. So, I thought it best to handle that case in code (even if it's unlikely that it will ever be executed). Hopefully, for most of the field that come along in the API we'll be able to get away with relatively simple regexs and lengths and we'll make use of this reconciler validation sparingly.

Since I'm rubber ducking this with you now, it occured to me that should a false positive slip through, we'd get a resolution failure on trying to parse the semver. So, maybe I'm being overzealous? Though a part of me thinks that ensuring that the input is clean before reaching the resolver might be worthwhile.

Reply:
It indeed feels like belt and braces, but I think I'm in favour of keeping this validation as it prevents us hittng rest of the code in case of discrapancies between the regex and the library (e.g. bugs in library, for example).

Reply:
Assuming you got the regex from a reliable source, this scenario should never be hit, but better safe than sorry.